src/container/srv_container.c

```diff
@@ -1647,196 +1649,248 @@ cont_destroy(struct rdb_tx *tx, struct ds_pool_hdl *pool_hdl, struct cont *cont,
+			rank, ec_agg_eph, stable_eph);
 	}
 
 out_put:
 	cont_svc_put_leader(svc);
 	return 0;
 }
 
-struct refresh_vos_agg_eph_arg {
-	uuid_t	pool_uuid;
-	uuid_t  cont_uuid;
-	daos_epoch_t min_eph;
+#define EPH_ARG_TGT_INLINE	(64)
+struct refresh_track_eph_arg {
+	uuid_t		 pool_uuid;
+	uuid_t		 cont_uuid;
+	daos_epoch_t	 min_ec_agg_eph;
+	daos_epoch_t	 min_stable_eph;
+	uint8_t		*tgt_status;
+	uint8_t		 tgt_status_inline[EPH_ARG_TGT_INLINE];
 };
 
-int
-cont_refresh_vos_agg_eph_one(void *data)
+static int
+cont_refresh_track_eph_one(void *data)
 {
-	struct refresh_vos_agg_eph_arg *arg = data;
-	struct ds_cont_child	*cont_child;
-	int			rc;
+	struct refresh_track_eph_arg	*arg = data;
+	struct ds_cont_child		*cont_child;
+	unsigned int			 idx = dss_get_module_info()->dmi_tgt_id;
+	int				 rc;
 
 	rc = ds_cont_child_lookup(arg->pool_uuid, arg->cont_uuid, &cont_child);
 	if (rc)
 		return rc;
 
-	D_DEBUG(DB_MD, DF_CONT": %s agg boundary eph "DF_X64"->"DF_X64"\n",
+	D_DEBUG(DB_MD, DF_CONT": %s ec agg boundary eph "DF_X64"->"DF_X64", "
+		": %s stable eph "DF_X64"->"DF_X64"\n",
 		DP_CONT(arg->pool_uuid, arg->cont_uuid),
-		cont_child->sc_ec_agg_eph_boundary < arg->min_eph ? "update" : "ignore",
-		cont_child->sc_ec_agg_eph_boundary, arg->min_eph);
-
-	if (cont_child->sc_ec_agg_eph_boundary < arg->min_eph)
-		cont_child->sc_ec_agg_eph_boundary = arg->min_eph;
+		cont_child->sc_ec_agg_eph_boundary < arg->min_ec_agg_eph ? "update" : "ignore",
+		cont_child->sc_ec_agg_eph_boundary, arg->min_ec_agg_eph,
+		cont_child->sc_stable_epoch < arg->min_stable_eph ? "update" : "ignore",
+		cont_child->sc_stable_epoch, arg->min_stable_eph);
+
+	if (cont_child->sc_ec_agg_eph_boundary < arg->min_ec_agg_eph)
+		cont_child->sc_ec_agg_eph_boundary = arg->min_ec_agg_eph;
+
+	/* Only should update local stable epoch if the target is in UPIN status */
+	if (cont_child->sc_stable_epoch < arg->min_stable_eph &&
+	    (arg->tgt_status[idx] & PO_COMP_ST_UPIN)) {
+		rc = vos_cont_set_global_stable_epoch(cont_child->sc_hdl, arg->min_stable_eph);
+		if (rc == 0)
+			cont_child->sc_stable_epoch = arg->min_stable_eph;
+		else
+			rc = 0;
+	}
 
 	ds_cont_child_put(cont_child);
 	return rc;
 }
 
 int
-ds_cont_tgt_refresh_agg_eph(uuid_t pool_uuid, uuid_t cont_uuid,
-			    daos_epoch_t eph)
+ds_cont_tgt_refresh_track_eph(uuid_t pool_uuid, uuid_t cont_uuid,
+			      daos_epoch_t ec_agg_eph, daos_epoch_t stable_eph)
 {
-	struct refresh_vos_agg_eph_arg	arg;
-	int				rc;
+	struct ds_pool			*pool;
+	struct pool_target		*tgts;
+	struct refresh_track_eph_arg	 arg;
+	d_rank_t			 rank;
+	int				 i, rc;
 
 	uuid_copy(arg.pool_uuid, pool_uuid);
 	uuid_copy(arg.cont_uuid, cont_uuid);
-	arg.min_eph = eph;
+	arg.min_ec_agg_eph = ec_agg_eph;
+	arg.min_stable_eph = stable_eph;
+	if (likely(dss_tgt_nr <= EPH_ARG_TGT_INLINE)) {
+		arg.tgt_status = arg.tgt_status_inline;
+	} else {
+		D_ALLOC_ARRAY(arg.tgt_status, dss_tgt_nr);
+		if (arg.tgt_status == NULL) {
+			ds_pool_put(pool);
```

Comment:
`pool` is NULL here.

Reply:
maybe an issue when rebase, will fix it soon, thx